File: ios/engine/BridgeKitNative.swift
// BridgeKitNative.swift
// BridgeKit iOS engine — singleton seam between Nitro Hybrid objects and the Swift core.
//
// `delegate` defaults to NotReadyDelegate (returns BRIDGE_NOT_READY) and is swapped
// once at startup. NSLock + nonisolated(unsafe) provide thread safety because Nitro
// callbacks arrive on arbitrary threads and the delegate must be settable synchronously
// (an actor would force async access).

import Foundation

// MARK: - BridgeKitNativeDelegate

/// Full delegate surface exposed to the Swift core.
/// All [String: Any?] values conform to the {v: <value>} wire rule.
public protocol BridgeKitNativeDelegate: AnyObject {

    // BridgeHost
    func invoke(env: [String: Any?], complete: @escaping ([String: Any?]) -> Void)
    func invokeSync(env: [String: Any?]) -> [String: Any?]
    func connectDispatcher(epochInfo: [String: Any?], callbacks: JsDispatcherCallbacks) -> [String: Any?]

    // BridgeStreams
    func openStream(env: [String: Any?], onNext: @escaping ([String: Any?]) -> Void, onEnd: @escaping ([String: Any?]) -> Void) -> String
    func closeStream(streamId: String)
    func emitFromJs(streamId: String, value: [String: Any?])
    func endFromJs(streamId: String, end: [String: Any?])

    // BridgeState
    func stateRead(env: [String: Any?]) -> [String: Any?]
    func stateObserve(env: [String: Any?], onChange: @escaping ([String: Any?]) -> Void) -> String
    func stateUnobserve(obsId: String)
    func stateWrite(env: [String: Any?]) -> [String: Any?]
}

// MARK: - BridgeKitNative singleton

/// Singleton seam — HybridBridgeHost/State/Streams delegate every call here.
public final class BridgeKitNative {

    /// Shared singleton.
    public static let shared = BridgeKitNative()

    // nonisolated(unsafe): set once before Nitro host creation; NSLock guards the window.
    private let _lock = NSLock()
    nonisolated(unsafe) private var _delegate: BridgeKitNativeDelegate = NotReadyDelegate.shared

    private init() {}

    /// Get/set the active delegate.
    ///
    /// Thread-safe: NSLock-protected. Set BEFORE Nitro host initialization.
    public var delegate: BridgeKitNativeDelegate {
        get {
            _lock.lock(); defer { _lock.unlock() }
            return _delegate
        }
        set {
            _lock.lock()
            _delegate = newValue
            _lock.unlock()
        }
    }
}

// MARK: - NotReadyDelegate

private let BRIDGE_NOT_READY_MSG = "BridgeKit native core not initialized. " +
    "Ensure the JS bundle entry imports '@malopezr7/bridgekit' before the first call."

private func notReadyEnvelope() -> [String: Any?] {
    ["ok": false, "code": "BRIDGE_NOT_READY", "message": BRIDGE_NOT_READY_MSG]
}

/// Default delegate — returns BRIDGE_NOT_READY envelopes until the real impl is wired in.
public final class NotReadyDelegate: BridgeKitNativeDelegate {

    public static let shared = NotReadyDelegate()
    private init() {}

    public func invoke(env: [String: Any?], complete: @escaping ([String: Any?]) -> Void) {
        complete(notReadyEnvelope())
    }

    public func invokeSync(env: [String: Any?]) -> [String: Any?] {
        notReadyEnvelope()
    }

    public func connectDispatcher(epochInfo: [String: Any?], callbacks: JsDispatcherCallbacks) -> [String: Any?] {
        ["epoch": 0, "snapshot": [] as [[String: Any?]]]
    }

    public func openStream(
        env: [String: Any?],
        onNext: @escaping ([String: Any?]) -> Void,
        onEnd: @escaping ([String: Any?]) -> Void
    ) -> String {
        onEnd(notReadyEnvelope())
        return ""
    }

    public func closeStream(streamId: String) {}
    public func emitFromJs(streamId: String, value: [String: Any?]) {}
    public func endFromJs(streamId: String, end: [String: Any?]) {}

    public func stateRead(env: [String: Any?]) -> [String: Any?] { notReadyEnvelope() }
    public func stateObserve(env: [String: Any?], onChange: @escaping ([String: Any?]) -> Void) -> String { "" }
    public func stateUnobserve(obsId: String) {}
    public func stateWrite(env: [String: Any?]) -> [String: Any?] { notReadyEnvelope() }
}

File: packages/core/android/src/main/java/com/bridgekit/runtime/BridgeKitNative.kt
package com.bridgekit.runtime

/**
 * Singleton seam between the Nitro Hybrid objects (Slice C) and the BridgeKit
 * Kotlin core (Slice D). HybridBridgeHost / HybridBridgeStreams / HybridBridgeState
 * delegate every call here.
 *
 * Defaults to [NotReadyDelegate] which returns BRIDGE_NOT_READY envelopes for
 * all calls and is a no-op for fire-and-forget paths.
 *
 * Slice D wires in a real implementation via:
 *   BridgeKitNative.delegate = realDelegate
 */
object BridgeKitNative {
    @Volatile
    var delegate: BridgeKitNativeDelegate = NotReadyDelegate
}

/**
 * Full delegate surface exposed to the Kotlin core (Slice D).
 * All Map<String, Any?> values conform to the { v: <value> } wire rule
 * documented in the BridgeHost / BridgeState specs.
 */
interface BridgeKitNativeDelegate {
    // ---- BridgeHost ----------------------------------------------------------

    /**
     * Async invoke. Calls [complete] exactly once with the result envelope map.
     */
    fun invoke(env: Map<String, Any?>, complete: (Map<String, Any?>) -> Unit)

    /**
     * Synchronous invoke — only lock-free in-memory reads are permitted.
     * Returns a result envelope map.
     */
    fun invokeSync(env: Map<String, Any?>): Map<String, Any?>

    /**
     * Register the JS dispatcher for a new runtime epoch.
     * Returns { epoch: Number, snapshot: List<Map<String,Any?>> } as a map.
     */
    fun connectDispatcher(
        epochInfo: Map<String, Any?>,
        callbacks: JsDispatcherCallbacks,
    ): Map<String, Any?>

    // ---- BridgeStreams -------------------------------------------------------

    /**
     * Open a native→JS stream. Returns the stream id (epoch-scoped).
     */
    fun openStream(
        env: Map<String, Any?>,
        onNext: (Map<String, Any?>) -> Unit,
        onEnd: (Map<String, Any?>) -> Unit,
    ): String

    /** Cancel a native→JS stream from the JS side. */
    fun closeStream(streamId: String)

    /**
     * Push a value from the JS producer to the Kotlin consumer.
     * value is wrapped { v: <encoded-value> }.
     */
    fun emitFromJs(streamId: String, value: Map<String, Any?>)

    /**
     * Signal end-of-stream from the JS producer.
     * end is a ResultEnvelope map.
     */
    fun endFromJs(streamId: String, end: Map<String, Any?>)

    // ---- BridgeState --------------------------------------------------------

    /** Synchronous state read. Returns a ResultEnvelope map. */
    fun stateRead(env: Map<String, Any?>): Map<String, Any?>

    /**
     * Subscribe to state changes. Returns obsId.
     * onChange receives { v: <encoded-value> }.
     */
    fun stateObserve(env: Map<String, Any?>, onChange: (Map<String, Any?>) -> Unit): String

    /** Cancel a state observation. */
    fun stateUnobserve(obsId: String)

    /**
     * Provider-side state write from JS.
     * Returns a ResultEnvelope map. Rejects with NOT_PROVIDER if the caller
     * doesn't own the binding.
     */
    fun stateWrite(env: Map<String, Any?>): Map<String, Any?>
}

/**
 * Callbacks that the Kotlin dispatcher passes back to JS.
 * The double-Promise from the Nitro callback (onInvoke) is encapsulated HERE:
 * [onInvoke] exposes a plain completion-callback API; HybridBridgeHost
 * adapts the Nitro Func_* + Promise plumbing.
 */
class JsDispatcherCallbacks(
    /**
     * Route a native→JS method call.
     * Call [completion] with (result, null) on success or (null, err) on failure.
     * Never throws — implementations must catch all exceptions.
     */
    val onInvoke: (
        env: Map<String, Any?>,
        completion: (ok: Map<String, Any?>?, err: Throwable?) -> Unit,
    ) -> Unit,

    /** Signal a JS-provided stream to start emitting. env carries the open envelope. */
    val onStreamOpen: (env: Map<String, Any?>) -> Unit,

    /**
     * Signal teardown of a native-side stream consumer.
     * env = { streamId: String, reason: String }.
     */
    val onStreamClose: (env: Map<String, Any?>) -> Unit,

    /** Push a state-change notification to JS mirrors. env is the StateWrite envelope. */
    val onStateWrite: (env: Map<String, Any?>) -> Unit,
)

// ---- NotReadyDelegate -------------------------------------------------------

private const val BRIDGE_NOT_READY_MSG = "BridgeKit native core not initialized. " +
    "Ensure the JS bundle entry imports '@malopezr7/bridgekit' before the first call."

private fun notReadyEnvelope(): Map<String, Any?> = mapOf(
    "ok" to false,
    "code" to "BRIDGE_NOT_READY",
    "message" to BRIDGE_NOT_READY_MSG,
)

/**
 * Default delegate — returns BRIDGE_NOT_READY envelopes until Slice D wires in the real impl.
 */
object NotReadyDelegate : BridgeKitNativeDelegate {
    override fun invoke(env: Map<String, Any?>, complete: (Map<String, Any?>) -> Unit) {
        complete(notReadyEnvelope())
    }

    override fun invokeSync(env: Map<String, Any?>): Map<String, Any?> = notReadyEnvelope()

    override fun connectDispatcher(
        epochInfo: Map<String, Any?>,
        callbacks: JsDispatcherCallbacks,
    ): Map<String, Any?> = mapOf(
        "epoch" to 0,
        "snapshot" to emptyList<Map<String, Any?>>(),
    )

    override fun openStream(
        env: Map<String, Any?>,
        onNext: (Map<String, Any?>) -> Unit,
        onEnd: (Map<String, Any?>) -> Unit,
    ): String {
        onEnd(notReadyEnvelope())
        return ""
    }

    override fun closeStream(streamId: String) { /* no-op */ }

    override fun emitFromJs(streamId: String, value: Map<String, Any?>) { /* no-op */ }

    override fun endFromJs(streamId: String, end: Map<String, Any?>) { /* no-op */ }

    override fun stateRead(env: Map<String, Any?>): Map<String, Any?> = notReadyEnvelope()

    override fun stateObserve(env: Map<String, Any?>, onChange: (Map<String, Any?>) -> Unit): String = ""

    override fun stateUnobserve(obsId: String) { /* no-op */ }

    override fun stateWrite(env: Map<String, Any?>): Map<String, Any?> = notReadyEnvelope()
}
